diff --git a/contrib/setup-btor2tools.sh b/contrib/setup-btor2tools.sh
index 524ed18a..398d8788 100755
--- a/contrib/setup-btor2tools.sh
+++ b/contrib/setup-btor2tools.sh
@@ -7,7 +7,7 @@ source "$(dirname "$0")/setup-utils.sh"
 BTOR2TOOLS_DIR=${DEPS_DIR}/btor2tools
 COMMIT_ID="1df768d75adfb13a8f922f5ffdd1d58e80cb1cc2"
 
-download_github "boolector/btor2tools" "$COMMIT_ID" "$BTOR2TOOLS_DIR"
+#download_github "boolector/btor2tools" "$COMMIT_ID" "$BTOR2TOOLS_DIR"
 cd "${BTOR2TOOLS_DIR}"
 
 if is_windows; then
@@ -16,7 +16,9 @@ if is_windows; then
   test_apply_patch "${component}" "${last_patch_date}"
 fi
 
-./configure.sh -fPIC
+./configure.sh --static
+cd build
 make -j${NPROC}
-install_lib build/libbtor2parser.a
+cd ..
+install_lib build/lib/libbtor2parser.a
 install_include src/btor2parser/btor2parser.h btor2parser
diff --git a/contrib/setup-cadical.sh b/contrib/setup-cadical.sh
index 23e82418..6e5a87ab 100755
--- a/contrib/setup-cadical.sh
+++ b/contrib/setup-cadical.sh
@@ -13,7 +13,7 @@ if is_windows; then
   TAR_ARGS="--exclude src/makefile"
 fi
 
-download_github "arminbiere/cadical" "$COMMIT_ID" "$CADICAL_DIR" "$TAR_ARGS"
+#download_github "arminbiere/cadical" "$COMMIT_ID" "$CADICAL_DIR" "$TAR_ARGS"
 cd "${CADICAL_DIR}"
 
 if is_windows; then
diff --git a/contrib/setup-lingeling.sh b/contrib/setup-lingeling.sh
index ac0b4bff..8e701300 100755
--- a/contrib/setup-lingeling.sh
+++ b/contrib/setup-lingeling.sh
@@ -7,7 +7,7 @@ source "$(dirname "$0")/setup-utils.sh"
 LINGELING_DIR=${DEPS_DIR}/lingeling
 COMMIT_ID="03b4860d14016f42213ea271014f2f13d181f504"
 
-download_github "arminbiere/lingeling" "$COMMIT_ID" "$LINGELING_DIR"
+#download_github "arminbiere/lingeling" "$COMMIT_ID" "$LINGELING_DIR"
 cd "${LINGELING_DIR}"
 
 if is_windows; then
diff --git a/contrib/setup-utils.sh b/contrib/setup-utils.sh
index b2431949..8c2a3045 100644
--- a/contrib/setup-utils.sh
+++ b/contrib/setup-utils.sh
@@ -10,8 +10,8 @@ die () {
 
 [ ! -e src/boolector.c ] && die "$0 not called from Boolector base directory"
 
-DEPS_DIR="$(pwd)/deps"
-INSTALL_DIR="${DEPS_DIR}/install"
+DEPS_DIR="${BUILD_DIR}"
+INSTALL_DIR="$(pwd)/deps/install"
 INSTALL_LIB_DIR="${INSTALL_DIR}/lib"
 INSTALL_INCLUDE_DIR="${INSTALL_DIR}/include"
 INSTALL_BIN_DIR="${INSTALL_DIR}/bin"
