diff --git a/src/expr/mkmetakind b/src/expr/mkmetakind
index 9d2f0a475..ea4b2fba7 100755
--- a/src/expr/mkmetakind
+++ b/src/expr/mkmetakind
@@ -16,6 +16,9 @@
 #
 # Output is to standard out.
 #
+# Required to disable this option for bash >=5.2 to avoid automatically
+# replacing & by the substituted text.
+shopt | grep -q '^patsub_replacement\b' &&  shopt -u patsub_replacement
 
 copyright=2010-2014
 
diff --git a/src/include/cvc4_public.h b/src/include/cvc4_public.h
index 9f6dbd753..7b17a08c3 100644
--- a/src/include/cvc4_public.h
+++ b/src/include/cvc4_public.h
@@ -21,7 +21,9 @@
 
 #include <stddef.h>
 #include <stdint.h>
-
+#if defined __cplusplus
+#include <cstdint>
+#endif
 #if defined _WIN32 || defined __CYGWIN__
 #  define CVC4_PUBLIC
 #else /* !( defined _WIN32 || defined __CYGWIN__ ) */
