diff --git a/configure.ac b/configure.ac
index e4714d7e..13585836 100644
--- a/configure.ac
+++ b/configure.ac
@@ -14,6 +14,7 @@ AC_CONFIG_HEADERS([tgt-pcb/pcb_config.h])
 AC_CANONICAL_HOST
 dnl Checks for programs.
 AC_PROG_CC
+AX_PROG_CC_FOR_BUILD
 AC_PREREQ([2.62])
 m4_version_prereq([2.70], [], [AC_PROG_CC_C99])
 AC_PROG_CXX
@@ -333,6 +334,5 @@ then
 AC_MSG_ERROR(cannot configure white space in libdir: $libdir)
 fi
 AC_MSG_RESULT(ok)
-AX_PROG_CC_FOR_BUILD
 AC_CONFIG_FILES([Makefile ivlpp/Makefile vhdlpp/Makefile vvp/Makefile vpi/Makefile driver/Makefile driver-vpi/Makefile cadpli/Makefile libveriuser/Makefile tgt-null/Makefile tgt-stub/Makefile tgt-vvp/Makefile tgt-vhdl/Makefile tgt-fpga/Makefile tgt-verilog/Makefile tgt-pal/Makefile tgt-vlog95/Makefile tgt-pcb/Makefile tgt-blif/Makefile tgt-sizer/Makefile])
 AC_OUTPUT
diff --git a/vvp/Makefile.in b/vvp/Makefile.in
index e8a944f6..3dbd57b2 100644
--- a/vvp/Makefile.in
+++ b/vvp/Makefile.in
@@ -146,7 +146,7 @@ endif
 	mv $*.d dep/$*.d
 
 tables.cc: $(srcdir)/draw_tt.c
-	$(CC) $(CFLAGS) -o draw_tt$(BUILDEXT) $(srcdir)/draw_tt.c
+	$(BUILDCC) $(CFLAGS) -o draw_tt$(BUILDEXT) $(srcdir)/draw_tt.c
 	./draw_tt$(BUILDEXT) > tables.cc
 	rm draw_tt$(BUILDEXT)
 
diff --git a/driver/main.c b/driver/main.c
index bce40319..34caf745 100644
--- a/driver/main.c
+++ b/driver/main.c
@@ -1053,7 +1053,12 @@ static void find_ivl_root(void)
 	/* Convert to a short name to remove any embedded spaces. */
       len = GetShortPathName(tmppath, ivl_root, sizeof ivl_root);
 #else
+	#ifdef __APPLE__
+	  uint32_t size = sizeof ivl_root;
+	  len = _NSGetExecutablePath(ivl_root, &size) + 1;
+	#else
       len = readlink("/proc/self/exe", ivl_root, sizeof ivl_root);
+	#endif
 #endif
       if (len >= (ssize_t) sizeof ivl_root) {
 	    find_ivl_root_failed("command path exceeds size of string buffer.");
