diff --git a/configure.ac b/configure.ac
index efe6922b5d..7c0982663f 100644
--- a/configure.ac
+++ b/configure.ac
@@ -337,6 +337,9 @@ then
 	*-*-linux*)
 		ac_sys_system=Linux
 		;;
+	*-*-darwin*)
+		ac_sys_system=Darwin
+		;;
 	*-*-cygwin*)
 		ac_sys_system=Cygwin
 		;;
@@ -383,6 +386,15 @@ if test "$cross_compiling" = yes; then
 			_host_cpu=$host_cpu
 		esac
 		;;
+	*-*-darwin*)
+		case "$host_cpu" in
+		arm*)
+			_host_cpu=arm
+			;;
+		*)
+			_host_cpu=$host_cpu
+		esac
+		;;
 	*-*-cygwin*)
 		_host_cpu=
 		;;
@@ -2046,33 +2058,7 @@ case $ac_sys_system/$ac_sys_release in
     }
     ]])],[ac_osx_32bit=yes],[ac_osx_32bit=no],[ac_osx_32bit=yes])
     
-    if test "${ac_osx_32bit}" = "yes"; then
-    	case `/usr/bin/arch` in
-    	i386) 
-    		MACOSX_DEFAULT_ARCH="i386" 
-    		;;
-    	ppc) 
-    		MACOSX_DEFAULT_ARCH="ppc" 
-    		;;
-    	*)
-    		AC_MSG_ERROR([Unexpected output of 'arch' on OSX])
-    		;;
-    	esac
-    else
-    	case `/usr/bin/arch` in
-    	i386) 
-    		MACOSX_DEFAULT_ARCH="x86_64" 
-    		;;
-    	ppc) 
-    		MACOSX_DEFAULT_ARCH="ppc64" 
-    		;;
-    	*)
-    		AC_MSG_ERROR([Unexpected output of 'arch' on OSX])
-    		;;
-    	esac
-
-    fi
-
+    MACOSX_DEFAULT_ARCH="x86_64" 
     LIBTOOL_CRUFT=$LIBTOOL_CRUFT" -lSystem -lSystemStubs -arch_only ${MACOSX_DEFAULT_ARCH}"
     LIBTOOL_CRUFT=$LIBTOOL_CRUFT' -install_name $(PYTHONFRAMEWORKINSTALLDIR)/Versions/$(VERSION)/$(PYTHONFRAMEWORK)'
     LIBTOOL_CRUFT=$LIBTOOL_CRUFT' -compatibility_version $(VERSION) -current_version $(VERSION)';;
